src/game/player_super_sonic.c

```diff
@@ -1,4 +1,867 @@
+    sSuperSonicTask = t;
+    sonic = TASK_DATA(t);
+
+    sonic->flags = SUPER_FLAG__20;
+    sonic->worldX = Q_24_8(600);
+    sonic->worldY = Q_24_8(288);
+    sonic->unk10 = 0;
+    sonic->unk14 = 0;
+    sonic->unk1A = 0;
+    sonic->rotation = 0;
+    sonic->unk20 = 0;
+    sonic->unkC = 300;
+    sonic->unk22 = 0;
+    sonic->func24 = sub_802C828;
+    sonic->unk128 = 0;
+    sonic->rawKeys = gInput;
+    sonic->pressedKeys = gPressedKeys;
+    sonic->shouldDestroy = FALSE;
+
+    spr = &sonic->spr;
+    spr->graphics.dest = RESERVED_SUPER_SONIC_TILES_VRAM;
+    spr->graphics.anim = gAnims_SuperSonic_080D69C8[0].anim;
+    spr->variant = gAnims_SuperSonic_080D69C8[0].variant;
+    spr->unk1A = SPRITE_OAM_ORDER(8);
+    spr->graphics.size = 0;
+    spr->x = 0;
+    spr->y = 0;
+    spr->animCursor = 0;
+    spr->timeUntilNextFrame = 0;
+    spr->prevVariant = -1;
+    spr->animSpeed = SPRITE_ANIM_SPEED(1.0);
+    spr->palId = 0;
+    spr->hitboxes[0].index = -1;
+    spr->hitboxes[1].index = -1;
+    spr->unk10 = (SPRITE_FLAG(PRIORITY, 1) | SPRITE_FLAG_MASK_X_FLIP);
+}
+
+void sub_802B81C(void)
+{
+    struct SuperSonic *sonic;
+
+    if (!sSuperSonicTask) {
+        return;
+    }
+
+    sonic = TASK_DATA(sSuperSonicTask);
+    sSuperSonicTask->main = Task_802C7E8;
+    sonic->func24 = sub_802C8A0;
+
+    SUPER_SWITCH_ANIM(sonic, 2);
+}
+
+static int sub_802B8A8(struct SuperSonic *sonic)
+{
+    int zero = 0;
+
+    if (!(sonic->flags & SUPER_FLAG__10) && !(gUnknown_03005424 & EXTRA_STATE__100)) {
+        if (gCourseTime >= MAX_COURSE_TIME && !gLoadedSaveGame->timeLimitDisabled) {
+            gPlayer.moveState = MOVESTATE_DEAD;
+
+            sonic->func24 = sub_802C8EC;
+            sonic->unkC = 60;
+            sonic->unk1A = 0;
+            sonic->rotation = 0x100;
+            sonic->flags = SUPER_FLAG__10;
+
+            SUPER_SWITCH_ANIM(sonic, 21);
+        } else if (!(sonic->flags & SUPER_FLAG__200)) {
+            if (Mod(gStageTime, GBA_FRAMES_PER_SECOND) == 0) {
+                if (gRingCount == 0) {
+                    u32 flags;
+                    gPlayer.moveState = MOVESTATE_DEAD;
+
+                    flags = SUPER_FLAG__10;
+
+                    sonic->func24 = sub_802C8EC;
+                    sonic->unkC = 60;
+                    sonic->unk1A = 0;
+                    sonic->rotation = Q_24_8(1.0);
+                    sonic->flags = flags;
+
+                    SUPER_SWITCH_ANIM(sonic, 21);
+
+                    m4aSongNumStart(SE_LIFE_LOST);
+                } else {
+                    gRingCount--;
+                }
+            }
+        } else {
+            if (ExtraBossIsDead() == TRUE) {
+                gUnknown_03005424 |= (EXTRA_STATE__ACT_START | EXTRA_STATE__2);
+                sonic->func24 = sub_802C8A0;
+
+                SUPER_SWITCH_ANIM(sonic, 2);
+            }
+        }
+    }
+
+#ifndef NON_MATCHING
+    asm("" ::"r"(zero));
+#endif
+
+    return 0;
```

Comment:
If you `return zero` does this not match?

Reply:
Nope, not last time I checked!